File: cv-assistant/src/search/fallback.py
from __future__ import annotations

from bs4 import BeautifulSoup
import requests

from search.models import SearchRequest, SearchSource
from search.utils import clean_text

DUCKDUCKGO_HTML_URL = "https://html.duckduckgo.com/html/"
SOURCE_HINTS: dict[SearchSource, str] = {
    "linkedin": "LinkedIn jobs",
    "seek": "SEEK jobs",
    "indeed": "Indeed jobs",
}
SOURCE_DOMAINS: dict[SearchSource, tuple[str, ...]] = {
    "linkedin": ("linkedin.com",),
    "seek": ("seek.com.au",),
    "indeed": ("indeed.com",),
}


def build_fallback_query(source: SearchSource, request: SearchRequest) -> str:
    return f'{request.jobTitle} {request.location} {SOURCE_HINTS[source]}'


def extract_duckduckgo_candidate_urls(source: SearchSource, html: str) -> list[str]:
    soup = BeautifulSoup(html, "html.parser")
    urls: list[str] = []

    for anchor in soup.find_all("a", href=True):
        href = clean_text(anchor.get("href"))
        if not href.startswith("http"):
            continue
        if not any(domain in href for domain in SOURCE_DOMAINS[source]):
            continue
        if href in urls:
            continue
        urls.append(href)

    return urls


def discover_fallback_urls(
    source: SearchSource,
    request: SearchRequest,
    session: requests.Session,
    limit: int = 3,
) -> list[str]:
    response = session.post(
        DUCKDUCKGO_HTML_URL,
        data={"q": build_fallback_query(source, request), "kl": "au-en"},
        timeout=20,
    )
    response.raise_for_status()
    return extract_duckduckgo_candidate_urls(source, response.text)[:limit]

File: cv-assistant/src/lib/search/server/fallback.ts
import { load } from "cheerio";
import type { SearchRequest, SearchSource } from "@/lib/search/types";
import { cleanText } from "@/lib/search/server/utils";

const DUCKDUCKGO_HTML_URL = "https://html.duckduckgo.com/html/";
const SOURCE_HINTS: Record<SearchSource, string> = {
  linkedin: "LinkedIn jobs",
  seek: "SEEK jobs",
  indeed: "Indeed jobs",
  careerone: "CareerOne jobs",
  adzuna: "Adzuna jobs",
  talent: "Talent.com jobs",
};

const SOURCE_DOMAINS: Record<SearchSource, string[]> = {
  linkedin: ["linkedin.com"],
  seek: ["seek.com.au"],
  indeed: ["indeed.com"],
  careerone: ["careerone.com.au"],
  adzuna: ["adzuna.com.au"],
  talent: ["talent.com"],
};

function buildFallbackQuery(source: SearchSource, request: SearchRequest) {
  return `${request.jobTitle} ${request.location} ${SOURCE_HINTS[source]}`;
}

export function extractDuckDuckGoCandidateUrls(source: SearchSource, html: string): string[] {
  const $ = load(html);
  const urls: string[] = [];

  $("a[href]").each((_, element) => {
    const href = cleanText($(element).attr("href"));
    if (!href.startsWith("http")) return;
    if (!SOURCE_DOMAINS[source].some((domain) => href.includes(domain))) return;
    if (urls.includes(href)) return;
    urls.push(href);
  });

  return urls;
}

export async function discoverFallbackUrls(
  source: SearchSource,
  request: SearchRequest,
  signal?: AbortSignal,
  limit = 3,
): Promise<string[]> {
  const response = await fetch(DUCKDUCKGO_HTML_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent":
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0 Safari/537.36",
      "Accept-Language": "en-AU,en;q=0.9",
    },
    body: new URLSearchParams({
      q: buildFallbackQuery(source, request),
      kl: "au-en",
    }),
    signal,
  });

  if (!response.ok) {
    throw new Error(`DuckDuckGo fallback failed with HTTP ${response.status}.`);
  }

  const html = await response.text();
  return extractDuckDuckGoCandidateUrls(source, html).slice(0, limit);
}
